File: backend/src/main/java/com/mysite/knitly/domain/community/comment/repository/CommentRepository.java
package com.mysite.knitly.domain.community.comment.repository;

import com.mysite.knitly.domain.community.comment.entity.Comment;
import com.mysite.knitly.domain.community.post.entity.Post;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.data.jpa.repository.*;
import org.springframework.data.repository.query.Param;

import java.util.List;
import java.util.Collection;


public interface CommentRepository extends JpaRepository<Comment, Long> {

    List<Comment> findByPost(Post post);

    // 정렬（등록순,최신순)
    Page<Comment> findByPostAndDeletedFalseOrderByCreatedAtAsc(Post post, Pageable pageable);
    Page<Comment> findByPostAndDeletedFalseOrderByCreatedAtDesc(Post post, Pageable pageable);

    // 댓글 수
    long countByPostIdAndDeletedFalse(Long postId);

    // 루트 댓글
    Page<Comment> findByPostAndParentIsNullAndDeletedFalseOrderByCreatedAtAsc(Post post, Pageable pageable);
    Page<Comment> findByPostAndParentIsNullAndDeletedFalseOrderByCreatedAtDesc(Post post, Pageable pageable);

    // 자식 대댓글
    List<Comment> findByParentIdAndDeletedFalseOrderByCreatedAtAsc(Long parentId);

    // 자식 대댓글 배치 조회 (N+1 제거)
    List<Comment> findByParentIdInAndDeletedFalseOrderByCreatedAtAsc(Collection<Long> parentIds);

    // 마이페이지 댓글 조회 시
    Page<Comment> findByAuthor_UserIdAndDeletedFalseAndContentContainingIgnoreCaseOrderByCreatedAtDesc(
            Long userId, String content, Pageable pageable
    );

    @Query("SELECT c.author.userId " +
            "FROM Comment c " +
            "WHERE c.deleted = false AND c.post.id = :postId " +
            "GROUP BY c.author.userId " +
            "ORDER BY MIN(c.createdAt) ASC")

    List<Long> findAuthorOrderForPost(@Param("postId") Long postId);

    List<Comment> findByPostIdAndDeletedFalseOrderByCreatedAtAsc(Long postId);
}

File: backend/src/main/kotlin/com/mysite/knitly/domain/community/comment/repository/CommentRepository.kt
package com.mysite.knitly.domain.community.comment.repository

import com.mysite.knitly.domain.community.comment.entity.Comment
import com.mysite.knitly.domain.community.post.entity.Post
import org.springframework.data.domain.Page
import org.springframework.data.domain.Pageable
import org.springframework.data.jpa.repository.JpaRepository
import org.springframework.data.jpa.repository.Query
import org.springframework.data.repository.query.Param

interface CommentRepository : JpaRepository<Comment, Long> {

    fun findByPost(post: Post): List<Comment>

    fun findByPostAndDeletedFalseOrderByCreatedAtAsc(post: Post, pageable: Pageable): Page<Comment>
    fun findByPostAndDeletedFalseOrderByCreatedAtDesc(post: Post, pageable: Pageable): Page<Comment>

    fun countByPostIdAndDeletedFalse(postId: Long): Long

    fun findByPostAndParentIsNullAndDeletedFalseOrderByCreatedAtAsc(post: Post, pageable: Pageable): Page<Comment>
    fun findByPostAndParentIsNullAndDeletedFalseOrderByCreatedAtDesc(post: Post, pageable: Pageable): Page<Comment>

    fun findByParentIdAndDeletedFalseOrderByCreatedAtAsc(parentId: Long): List<Comment>

    fun findByParentIdInAndDeletedFalseOrderByCreatedAtAsc(parentIds: Collection<Long>): List<Comment>

    fun findByAuthor_UserIdAndDeletedFalseAndContentContainingIgnoreCaseOrderByCreatedAtDesc(
        userId: Long,
        content: String,
        pageable: Pageable
    ): Page<Comment>

    @Query(
        "SELECT c.author.userId " +
                "FROM Comment c " +
                "WHERE c.deleted = false AND c.post.id = :postId " +
                "GROUP BY c.author.userId " +
                "ORDER BY MIN(c.createdAt) ASC"
    )
    fun findAuthorOrderForPost(@Param("postId") postId: Long): List<Long>

    fun findByPostIdAndDeletedFalseOrderByCreatedAtAsc(postId: Long): List<Comment>
}
